### PR TITLE
Add StateKeyTuple type. Add helper functions for auth.

### PR DESCRIPTION
**Pull request by Kegsay**
Helper functions are designed to map easily between StateNeeded,
StateKeyTuples and []EventReference (auth_events).

### CODE REVIEW

File: eventauth.go [hunk 1]
```diff
@@ -86,6 +86,40 @@ func StateNeededForAuth(events []Event) (result StateNeeded) {
...
+}
+
+// AuthEventReferencesFromStateNeeded returns a list of event references which can be used under the 'auth_events' key
+// of a matrix event. 'authEvents' should contain all the current state events which may be used for auth, as this
+// function may require their event reference.
+func AuthEventReferencesFromStateNeeded(authEvents map[StateKeyTuple]*Event, needed StateNeeded) (refs []EventReference) {
+	tuples := TuplesFromStateNeeded(needed)
+	for _, t := range tuples {
+		e, ok := authEvents[t]
+		if ok && e != nil {
```
Comment (NegativeMjark) [inline eventauth.go]: Technically the `ok` check is unnecessary here because `e` will be `nil` if the tuple isn't present in the map.

Comment (Kegsay) [inline eventauth.go]: Done. https://github.com/matrix-org/gomatrixserverlib/pull/28/commits/56239aa496b880a7ee4b7f4ee11cc72ffb4b70a3

File: eventauth.go [hunk 2]
```diff
@@ -86,6 +86,40 @@ func StateNeededForAuth(events []Event) (result StateNeeded) {
...
+	for _, token := range needed.ThirdPartyInvite {
+		res = append(res, StateKeyTuple{"m.room.third_party_invite", token})
+	}
+	return
+}
+
+// AuthEventReferencesFromStateNeeded returns a list of event references which can be used under the 'auth_events' key
+// of a matrix event. 'authEvents' should contain all the current state events which may be used for auth, as this
+// function may require their event reference.
+func AuthEventReferencesFromStateNeeded(authEvents map[StateKeyTuple]*Event, needed StateNeeded) (refs []EventReference) {
```
Comment (NegativeMjark) [inline eventauth.go]: I'd be happier if this was a method on `StateNeeded` that took a `AuthEventsProvider` interface.

File: eventauth.go [hunk 3]
```diff
@@ -86,6 +86,40 @@ func StateNeededForAuth(events []Event) (result StateNeeded) {
 	return
 }
 
+// TuplesFromStateNeeded returns the needed state key tuples for performing auth on an event.
+func TuplesFromStateNeeded(needed StateNeeded) (res []StateKeyTuple) {
```
Comment (NegativeMjark) [inline eventauth.go]: It might be neater to make this a method on `StateNeeded`.

File: eventauth.go (hunk 2 above)
Comment (NegativeMjark) [inline eventauth.go]: Something like this maybe?:

```go
func (needed StateNeeded) AuthEventReferences(authEvents AuthEventProvider) (refs []EventReference, err error) {
   var *event Event
   if needed.Create {
      if event, err = authEvents.Create(); err != nil {
           return
      } else if (event != nil) {
           append(refs, event.EventReference())
      }
   }
   ...
}
```

File: eventauth.go (hunk 3 above)
Comment (Kegsay) [inline eventauth.go]: Done. https://github.com/matrix-org/gomatrixserverlib/pull/28/commits/93b606350db0e699bd9ab4be57d158a9cdddce72

File: eventauth.go (hunk 2 above)
Comment (Kegsay) [inline eventauth.go]: SGTM. Done. https://github.com/matrix-org/gomatrixserverlib/pull/28/commits/93b606350db0e699bd9ab4be57d158a9cdddce72

File: eventauth.go [hunk 4]
```diff
@@ -44,6 +44,69 @@ type StateNeeded struct {
...
+	}
+	for _, token := range s.ThirdPartyInvite {
+		res = append(res, StateKeyTuple{"m.room.third_party_invite", token})
+	}
+	return
+}
+
+// AuthEventReferences returns the auth_events references for the StateNeeded. Returns an error if the
+// provider returns an error. If an event is missing from the provider but is required in StateNeeded, it
+// is skipped under the assumption that the event will fail authentication checks.
```
Comment (NegativeMjark) [inline eventauth.go]: Strictly speaking this is misleading. An event may still pass the authentication checks even though a tuple is missing from StateNeeded. For example `m.room.member` events need the previous value of the `m.room.member` event to auth, but when a user joins the room for the first time there is no previous event.

Comment (Kegsay) [inline eventauth.go]: Tweaked comment: https://github.com/matrix-org/gomatrixserverlib/pull/28/commits/121fba145760eda2b58e95604f0e9e720b08da5f

**Pull request merged**